### ISSUE TITLE
Source map and offset converter is not loaded or created when providing a custom instantiateWasm function

### ISSUE THREAD

**Issue by jgranstrom** [emscripten-core/emscripten#9836]
If we provide a custom function for downloading and instantiating the wasm binary we are not getting source map loaded or offset converter created, thus getting stuck at waiting for those runtime dependencies if for example enabling sanitizer.

It looks like when providing this it bails out early https://github.com/emscripten-core/emscripten/blob/6df46b605078ba132821facbfc6d8946c838227e/src/preamble.js#L1094-L1105 and never loads source map which for async is later and for sync is in the instantiator itself. https://github.com/emscripten-core/emscripten/blob/6df46b605078ba132821facbfc6d8946c838227e/src/preamble.js#L1113
https://github.com/emscripten-core/emscripten/blob/6df46b605078ba132821facbfc6d8946c838227e/src/preamble.js#L1086

In addition the offset converter seems to be created only from the built-in instantiations at https://github.com/emscripten-core/emscripten/blob/6df46b605078ba132821facbfc6d8946c838227e/src/preamble.js#L1036-L1047 https://github.com/emscripten-core/emscripten/blob/6df46b605078ba132821facbfc6d8946c838227e/src/preamble.js#L1064-L1073 
so these are also skipped when providing the custom function.

Is this intended, or is there a recommended way to go about this currently? There seems to be a few hoops to jump through to set this up from a custom instantiator properly.

**Comment by kripken** [comment]
I don't think this is intended, just that the source map support in the sanitizers has not been tested with custom `instantiateWasm`.

I actually forgot we allow custom `instantiateWasm` - why is it important for you?

**Comment by jgranstrom** [comment]
@kripken got it, we use it mainly for two reasons:

1. The binary location is determine at runtime with explicit version and bit version in the path (we can probably use `locateFile` for this).
2. Detailed error and progress reporting in order to make it easier for us to understand at what stage problems occur since we know at what stage the instantiation is currently at, and we control the network request. Mainly to avoid mapping the possible errors to what stage seem to be failing and speed up debugging production errors as well as presenting users with enough details to make support tickets originating from users more useful for such problems.

I think it should be possible to support this use case by allowing us to pass the binary back from the custom function for the offset converter and for preamble to load source map if this is requested after executing the custom function prior to bailing out. Exposing these steps to the custom instantiator to do manually would probably be quite error prone and wouldn't be guaranteed to conform to the flags provided at build time.

If not I think there is need for some note in the documentation, and possibly deprecating this ability altogether if it is likely that it will not have the same feature set as when not providing it. There might of course be other uses of `instantiateWasm` in the wild that can be more difficult to work around should it be removed.

**Comment by kripken** [comment]
Thanks @jgranstrom, interesting!

For 1, yeah, locateFile should be enough. 2 is the bigger issue I guess. I wonder if we can use a different type of hook for that - remove `instantiateWasm` and provide a hook specifically for error reporting?

I agree that meanwhile we should update the docs to mention this issue. Looks like the right file for that is `site/source/docs/api_reference/module.rst`. Maybe you want to open a PR?

**Comment by jgranstrom** [comment]
@kripken I opened a PR with a note in the docs on this https://github.com/emscripten-core/emscripten/pull/9900

I think what could replace instantiateWasm in order to allow users to tap into instantiation while still conforming to the standard instantiation procedure would be to provide some kind of hook or hooks  for error and progress reporting as you mentioned, but also to allow the user to create, or access, the network request in order to hook into more details as necessary, such as in our case:
```
wasmXHR.open('GET', url, true)
wasmXHR.responseType = 'arraybuffer'
wasmXHR.onload = (_) => /* custom things */
wasmXHR.onerror = (_) => /* custom things */ 
wasmXHR.onprogress = (e) =>  /* custom things */
wasmXHR.onloadend = (e) =>  /* custom things */
```

It seems like the general problem is that the scope of instantiateWasm is a bit too broad, making it problematic when emscripten have additional things to do that interleaves with what instantiateWasm is supposed to be doing, so breaking it up into smaller more specific hooks or features would probably be the way to go here.

So instead of the current configuration where you leave instantiation to the user:
```
Module = {
  instantiateWasm: () => {
    downloadBinary().then(binary => WebAssembly.instantiate(binary))
  }
}
```
You could keep instantiation internal and expose some more specific hooks instead such as:
```
Module = {
  fetchWasmBinary: () => downloadBinary(),
  onInstantiateError: (e) => /* custom handling */
}
```
So that emscripten is free do to additional work in between as necessary such as source map, offset converter and so on.

This could also allow for more hooks to cover other use cases if necessary, such as a hook to transform imports prior to passing them to WebAssembly.instantiate, which is just an example but to indicate that it could cover more use cases without having to expose the entirety of instantiateWasm.

**Comment by kripken** [comment]
I see @jgranstrom , interesting!

I'm not sure what is best here. On the one hand, we try to avoid adding multiple hooks, as each one has a cost in code size and complexity. On the other, those do seem like valid use cases. Maybe leaving the current hook is the least evil?

**Comment by jgranstrom** [comment]
@kripken hmm, yes it sure would be preferable to have as few hooks as possible. However leaving the current hook is problematic as well especially since even though right now the only affected features are offset converter and source map, but maybe there will be new additions that will have a similar fate where the usage of instantiateWasm would cause it to not function properly since those would probably be added similarly in instantiateSync/Async respectively.

I think the worst problem with leaving that hook is that it is that the error manifests in strange ways, since it just hangs during instantiation and from a user perspective you pretty much have to step through the preamble code to understand what is going wrong which can be quite time consuming. That could maybe be solved by keeping a list of unsupported runtime dependencies when using instantiateWasm and throw immediately with a descriptive error if those are in use when the user provided an instantiateWasm hook, there is always a risk that it would not be kept in sync of course.

**Comment by kripken** [comment]
Good point @jgranstrom, that's a definite risk. We should try to make features that don't work with it have assertions at least, but yes, it's very possible we will forget.

**Comment by jgranstrom** [comment]
@kripken maybe we can try to collect some other uses of `instantiateWasm` to see if it is even worth keeping? Since error handling seems like it would be an obvious thing that is useful for other users to have available, while the nature of `instantiateWasm` probably limits the use case of that specific hook as long as there is `locateFile` and some form of `onError`-ish hook replacing it. I would expect that removing it would simplify things in the preamble going forward. Maybe hooking into the network request itself is very specific to us, on the other hand it seems like it wouldn't be too crazy if others would also like to display similar progress reports to their users in case of significantly sized WASM binaries. 

Do you agree that more specific "smaller" hooks would be easier to maintain than the relatively big `instantiateWasm` one?

**Comment by kripken** [comment]
Collecting more use cases might be interesting. Perhaps the mailing list is a good place to ask, if you want to post there.

I'm not sure more small hooks would be simpler to maintain. My guess is that a single hook like we have now is simpler, just because of the overhead and breakage risk of adding any hook?

**Comment by dschuff** [comment]
This seems related to https://bugs.chromium.org/p/chromium/issues/detail?id=1016585 (and more generally the lack of progress events when using streaming instantiation is a regression that other folks coming from PNaCl have noticed).

**Comment by justjake** [comment]
Hello, I have a use-case for `instantiateWasm`. In Cloudflare Workers, the user isn't allowed to create a WebAssembly.Module from an ArrayBuffer, only the system can create WebAssembly.Module. There's no way to work around this limitation with `locateFile` - we need to inject WebAssembly.Module instance. 

When trying to inject `instantiateWasm`, I'm encountering a few issues:

- The hang on the source-map dependency never getting resolved.
- `wasmOffsetConverter` is never instantiated, so many assertions fail.

So far I'm working around these issues by careful coordination between my caller script and a `--pre-js`.

```javascript
// My pre-js
/* eslint-disable no-undef */
// quickjs-emscripten code injected into emscripten-module.js
// We use this to expose and patch up issues with Emscripten's source map handling...
function quickjsEmscriptenInit() {
  // Everything goes in a function so we can defer running until other variables
  // are initialized in case they change.
  const quickJSEmscriptenExtensions = {}
  if (typeof removeRunDependency !== 'undefined') {
    quickJSEmscriptenExtensions['removeRunDependency'] = removeRunDependency
  }
  if (typeof receiveSourceMapJSON !== 'undefined') {
    quickJSEmscriptenExtensions['receiveSourceMapJSON'] = receiveSourceMapJSON
  }
  if (typeof WasmOffsetConverter !== 'undefined') {
    quickJSEmscriptenExtensions['WasmOffsetConverter'] = WasmOffsetConverter
    // Expose function to receive WasmOffsetConverter, set to wasmOffsetConverter local variable
    // if it exists
    try {
      // Check if wasmOffsetConverter variable exists. If it isn't defined, this
      // will throw and we'll skip the rest of the branch.
      quickJSEmscriptenExtensions['existingWasmOffsetConverter'] = wasmOffsetConverter
      quickJSEmscriptenExtensions['canReceiveWasmOffsetConverter'] = true
      quickJSEmscriptenExtensions['receiveWasmOffsetConverter'] = function (wasmBinary, wasmModule) {
        wasmOffsetConverter = new WasmOffsetConverter(wasmBinary, wasmModule)
      }
    } catch (error) {
      // Nothing.
      quickJSEmscriptenExtensions['canReceiveWasmOffsetConverter'] = false
      quickJSEmscriptenExtensions['receiveWasmOffsetConverter'] = function () {
        console.warn('quickjs-emscripten: wasmOffsetConverter variable not defined, this is a no-op')
      }
    }
  }



  Module['quickJSEmscriptenExtensions'] = quickJSEmscriptenExtensions
  return quickJSEmscriptenExtensions
}

Module['quickjsEmscriptenInit'] = quickjsEmscriptenInit
```

Then in my caller that instantiates the WebAssembly module, i can poke these extensions shims:

```typescript
// My caller
const moduleLoader = import('./emscripten-module.mjs').default;
const moduleLoaderArg = {
	async instantiateWasm(imports: any, onSuccess: any) {
		const instance = await WebAssembly.instantiate(cloudflareWasmModule, imports);
		onSuccess(instance);
		return instance.exports;
	},
	locateFile(fileName: string) {
		if (fileName.endsWith('.map')) {
			return emptySourceMapDataURI;
		}
		return '';
	},
};
const resultPromise = moduleLoader(moduleLoaderArg);
const extensions = moduleLoaderArg.quickjsEmscriptenInit();
console.log('got extensions:', extensions);
console.log('runDependenciesLeft:', runDependenciesLeft);
if (extensions.WasmOffsetConverter) {
	extensions.receiveWasmOffsetConverter(new Uint8Array(0), cloudflareWasmModule);
}
extensions.receiveSourceMapJSON(emptySourceMapData);
return resultPromise;```

**Comment by justjake** [comment]
I also need to add this because WORKER environment (which I want to support as I'm building a library) incorrectly detects Cloudflare Worker as a WORKER, even though Cloudflare Worker doesn't have `globalThis.self`.
```js
// ENVIRONMENT_IS_WORKER checks `typeof importScripts == 'function'`
// then it assumes `self` is a global variable, but Cloudflare Workers
// don't have a global `self`, so a missing `self` implies we aren't a worker
if (typeof ENVIRONMENT_IS_WORKER && ENVIRONMENT_IS_WORKER && typeof self === 'undefined') {
  ENVIRONMENT_IS_WORKER = false
}
```
